File: esports-backend/src/main/java/com/esports/controller/FreelancerController.java
package com.esports.controller;

import com.esports.entity.Service;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

@RestController
@RequestMapping("/api/freelancer")
@CrossOrigin(origins = "http://localhost:5173")
@PreAuthorize("hasRole('FREELANCER') or hasRole('ADMIN')")
public class FreelancerController {

    @Autowired
    private FreelancerService freelancerService;

    @PostMapping(value = "/services", consumes = "multipart/form-data")
    public ResponseEntity<?> createService(
            @RequestParam("title") String title,
            @RequestParam("description") String description,
            @RequestParam("price") Double price,
            @RequestParam("deliveryTime") Integer deliveryTime,
            @RequestParam("category") String category,
            @RequestParam("active") Boolean active,
            @RequestParam(value = "photo1", required = false) MultipartFile photo1,
            @RequestParam(value = "photo2", required = false) MultipartFile photo2,
            @RequestParam(value = "photo3", required = false) MultipartFile photo3,
            Authentication auth) {

        try {
            Service service = freelancerService.createService(
                auth.getName(), title, description, price, deliveryTime,
                category, active, photo1, photo2, photo3
            );
            return ResponseEntity.ok(service);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "User not found"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Service creation failed"));
        }
    }

    @GetMapping("/services")
    public ResponseEntity<?> getFreelancerServices(Authentication auth) {
        try {
            List<ServiceResponse> services = freelancerService.getFreelancerServices(auth.getName());
            return ResponseEntity.ok(services);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "User not found"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to fetch services"));
        }
    }

    @PutMapping(value = "/services/{id}", consumes = "multipart/form-data")
    public ResponseEntity<?> updateService(
            @PathVariable Long id,
            @RequestParam("title") String title,
            @RequestParam("description") String description,
            @RequestParam("price") Double price,
            @RequestParam("deliveryTime") Integer deliveryTime,
            @RequestParam("category") String category,
            @RequestParam("active") Boolean active,
            @RequestParam(value = "photo1", required = false) MultipartFile photo1,
            @RequestParam(value = "photo2", required = false) MultipartFile photo2,
            @RequestParam(value = "photo3", required = false) MultipartFile photo3,
            Authentication auth) {

        try {
            Service service = freelancerService.updateService(
                auth.getName(), id, title, description, price, deliveryTime,
                category, active, photo1, photo2, photo3
            );
            return ResponseEntity.ok(Map.of(
                "message", "Service updated successfully",
                "service", service
            ));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Service or user not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Unauthorized"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to update service"));
        }
    }

    @GetMapping("/services/{id}")
    public ResponseEntity<?> getService(@PathVariable Long id) {
        try {
            Service service = freelancerService.getService(id);
            return ResponseEntity.ok(service);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Service not found"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to fetch service"));
        }
    }

    @DeleteMapping("/services/{id}")
    public ResponseEntity<?> deleteService(@PathVariable Long id, Authentication auth) {
        try {
            freelancerService.deleteService(auth.getName(), id);
            return ResponseEntity.ok(Map.of("message", "Service deleted successfully"));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Service or user not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Unauthorized"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to delete service"));
        }
    }
}

File: Upkart1/Controllers/FreelancerController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using Upkart1.Services;
using Upkart1.Filters;

namespace Upkart1.Controllers
{
    [ApiController]
    [Route("api/freelancer")]
    [Authorize]
    public class FreelancerController : ControllerBase
    {
        private readonly IFreelancerService _freelancerService;

        public FreelancerController(IFreelancerService freelancerService)
        {
            _freelancerService = freelancerService;
        }

        [HttpPost("services")]
        [Consumes("multipart/form-data")]
        public async Task<IActionResult> CreateService(
            [FromForm] string title,
            [FromForm] string description,
            [FromForm] double price,
            [FromForm] int deliveryTime,
            [FromForm] string category,
            [FromForm] bool active,
            [FromForm] IFormFile? photo1,
            [FromForm] IFormFile? photo2,
            [FromForm] IFormFile? photo3)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                Console.WriteLine($"Creating service for email: {email}");
                var service = await _freelancerService.CreateService(
                    email, title, description, price, deliveryTime,
                    category, active, photo1, photo2, photo3
                );
                return Ok(service);
            }
            catch (KeyNotFoundException ex)
            {
                Console.WriteLine($"User not found: {ex.Message}");
                return NotFound(new Dictionary<string, string> { { "error", "User not found" } });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Service creation error: {ex.Message}");
                Console.WriteLine($"Stack trace: {ex.StackTrace}");
                return StatusCode(500, new Dictionary<string, string> { { "error", "Service creation failed" } });
            }
        }

        [HttpGet("services")]
        public IActionResult GetFreelancerServices()
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var services = _freelancerService.GetFreelancerServices(email);
                return Ok(services);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new Dictionary<string, string> { { "error", "User not found" } });
            }
            catch (Exception)
            {
                return StatusCode(500, new Dictionary<string, string> { { "error", "Failed to fetch services" } });
            }
        }

        [HttpPut("services/{id}")]
        [Consumes("multipart/form-data")]
        public async Task<IActionResult> UpdateService(
            long id,
            [FromForm] string title,
            [FromForm] string description,
            [FromForm] double price,
            [FromForm] int deliveryTime,
            [FromForm] string category,
            [FromForm] bool active,
            [FromForm] IFormFile? photo1,
            [FromForm] IFormFile? photo2,
            [FromForm] IFormFile? photo3)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var service = await _freelancerService.UpdateService(
                    email, id, title, description, price, deliveryTime,
                    category, active, photo1, photo2, photo3
                );
                return Ok(new Dictionary<string, object>
                {
                    { "message", "Service updated successfully" },
                    { "service", service }
                });
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new Dictionary<string, string> { { "error", "Service or user not found" } });
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, new Dictionary<string, string> { { "error", "Unauthorized" } });
            }
            catch (Exception)
            {
                return StatusCode(500, new Dictionary<string, string> { { "error", "Failed to update service" } });
            }
        }

        [HttpGet("services/{id}")]
        public IActionResult GetService(long id)
        {
            try
            {
                var service = _freelancerService.GetService(id);
                return Ok(service);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new Dictionary<string, string> { { "error", "Service not found" } });
            }
            catch (Exception)
            {
                return StatusCode(500, new Dictionary<string, string> { { "error", "Failed to fetch service" } });
            }
        }

        [HttpDelete("services/{id}")]
        public IActionResult DeleteService(long id)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                _freelancerService.DeleteService(email, id);
                return Ok(new Dictionary<string, string> { { "message", "Service deleted successfully" } });
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new Dictionary<string, string> { { "error", "Service or user not found" } });
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, new Dictionary<string, string> { { "error", "Unauthorized" } });
            }
            catch (Exception)
            {
                return StatusCode(500, new Dictionary<string, string> { { "error", "Failed to delete service" } });
            }
        }
    }
}
